Getting Started Guide: Linux

MathiasMagnus opened the pull request:
This is a getting started guide focusing on Ubuntu, the distro of choice for Linux.

Kerilk commented:
I just noticed while reviewing the CI logs, the new file's extension is repeated.

ceich commented on chapters/getting_started_linux.md:
```diff
@@ -0,0 +1,308 @@
+
+### Visual Studio Code
+
+While IDEs are highly opinionated, for an IDE with widespared adoption and small footprint, VS Code has everything (and more) to get off the ground. It should be installed differently, though. The first way is to use GUI of `Ubuntu Software`, where you can search for `code` and install it. Whether you are a fan of command-line tools, use the second method: open your favorite shell in your favorit ...
```
The English of this paragraph needs some help. Here are my suggestions:
```suggestion
While IDEs are highly opinionated, for an IDE with widespread adoption and small footprint, VS Code has everything (and more) to get off the ground. It can be installed in two ways. The first way is to use the `Ubuntu Software` GUI, where you can search for `code` and install it in one click. If you are a fan of command-line tools, use the second method: open your favorite shell in your favorite terminal and run
```

bashbaug commented on chapters/getting_started_linux.md:
```diff
@@ -0,0 +1,308 @@
+  - The C and optionally the C++ headers
+- An Installable Client Driver (ICD) Loader
+  - Shared objects library (libOpenCL.so)
+  - Export library (libOpenCL.a)
```
On Linux a separate export library is not needed and the same libOpenCL.so is used for both purposes.

MathiasMagnus replied:
Yes, that was a mistake from our part. Fixed.

bashbaug commented on chapters/getting_started_linux.md:
````diff
@@ -0,0 +1,308 @@
+
+This will install canonical GNU compiler set `g++/GNU 9.3`, debugger `gdb`, as well as some tools (e.g. `make`, `binutils`) and libraries for compilation and linking of programs.
+
+Another popular choice is the usage of `Clang` compiler, based on intermediate representation of the C-like languages on `LLVM` (originally Low Level Virtual Machine). To use it open your favorite shell in your favorite terminal and run:
+
+``` bash
+sudo apt install clang -y
+```
+
+This will install `Clang 10` and `LLVM 10`.
````
I'd consider removing this section (and any other mention of Clang) entirely.  While I'm a fan of Clang, I think this is unnecessary to create a "minimal and productive environment" and as far as I can tell no part of this guide uses Clang.

bashbaug commented on chapters/getting_started_linux.md:
````diff
@@ -0,0 +1,308 @@
+If you do not have CMake installed yet, we'll install it, as it's the primary supported build system, and it will make our builds much simpler. Open your favorite shell in your favorite terminal and run:
+
+``` bash
+sudo apt install cmake ninja-build -y
````
Note, no part of this guide uses ninja.

MathiasMagnus replied:
Removed.

bashbaug reviewed (commented):
This is a nice guide, but I wonder if it is a little too complicated for a "getting started" guide.

An OpenCL fan has been contributing basic "setup" docs to my SimpleOpenCLSamples repo for a while now.  Here's an example for Ubuntu 18.04:

https://github.com/bashbaug/SimpleOpenCLSamples/blob/master/docs/env/ubuntu/18.04.md

It heavily leverages the Ubuntu packages for OpenCL headers / libs / etc.  These are admittedly a bit old, but they have the advantage of being super easy for novice users to install and are usually sufficient to build and run "my first OpenCL application".

How do you feel about having one "getting started" doc that uses OS packages to get up and running as quickly and easily as possible, and a different "getting start" doc like this one that generates an up-to-date environment built on the SDK?  They could even be two different sections of the same doc, if desired.

MathiasMagnus commented:
@bashbaug I added dropdown menus to the various ways of obtaining the headers/lib and how to build on the command-line in each case, but left uncollapsed by default only the the simplest way: APT. I left GitHub and Vcpkg there for those interested in those ways as well, but they're somewhat "out of sight, out of mind" and don't clutter the straightforward way.

Does that seem like something acceptable?

I also went ahead and removed the VS Code part because it was way too brief compared to the depth, so I will write a separate document which we can link from the end of both the Windows and Linux guide some other time.

bashbaug commented on chapters/getting_started_linux.md:
```diff
@@ -0,0 +1,324 @@
+In this guide we're minimally going to use the following tools:
+
+- The command-line
+- C/C++ compiler (gcc & Clang)
```
Minor: since we don't say anything about Clang anymore we may want to simplify this:
```suggestion
- C/C++ compiler
```

bashbaug commented on chapters/getting_started_linux.md:
```diff
@@ -0,0 +1,324 @@
+# Getting started with OpenCL on Ubuntu Linux
+
+OpenCL is included in all the main distributions of Linux. In this guide we choose Ubuntu as one of the most widespread variants, but all of the distributives have some sort of centralized system of package maintaining and OpenCL realizations can be installed from there.
```
```suggestion
OpenCL is included in all the main distributions of Linux. In this guide we choose Ubuntu as one of the most widespread variants, but all of the distributions have some sort of centralized system of package maintaining and OpenCL realizations can be installed from there.
```

bashbaug commented on chapters/getting_started_linux.md:
```diff
@@ -0,0 +1,324 @@
+
+## Installation
+
+In this guide we'll be using latest (at the time of writing) Ubuntu 20.04 LTS. Installation for the most part will happen via [apt (Advanced Packaging Tool)](https://en.wikipedia.org/wiki/APT_(software)), the definitive command-line tool for installing software on Debian-base Linux distributives. It is installed with the system automatically.
```
```suggestion
In this guide we'll be using latest (at the time of writing) Ubuntu 20.04 LTS. Installation for the most part will happen via [apt (Advanced Packaging Tool)](https://en.wikipedia.org/wiki/APT_(software)), the definitive command-line tool for installing software on Debian-base Linux distribution. It is installed with the system automatically.
```

bashbaug reviewed (approved):
This is a nice improvement!  Thank you for all of your work to improve the SDK!
